File: src/formatters/irs/pt/dividends_formatter.py
from src.models import Statement, Transaction

class PTDividendsFormatter:
    def __init__(self):
        pass
    
    def format(self, statement : Statement, year, currency = "EUR") -> list:
        # Obter transacoes referentes aos dividendos do ano a declarar
        transactions = statement.get_transactions()
        dividend_transactions = [transaction for transaction in transactions if transaction.type == "Dividend" and transaction.date[:4] == year]
        dividends = []
        dividends_by_asset = {}
        
        # Agrupar dividendos por ativo
        for dividend_transaction in dividend_transactions:
            if dividend_transaction.asset.isin not in dividends_by_asset:
                dividends_by_asset[dividend_transaction.asset.isin] = []
            dividends_by_asset[dividend_transaction.asset.isin].append(dividend_transaction)
        
        
        
        for _, dividend_transactions in dividends_by_asset.items():
            total_net_amount = sum([dividend_transaction.net_amount for dividend_transaction in dividend_transactions])
            total_net_amount = 0
            for dividend_transaction in dividend_transactions:
                if dividend_transaction.net_amount_currency == currency:
                    total_net_amount += dividend_transaction.net_amount
            total_fees_amount = 0
            for dividend_transaction in dividend_transactions:
                for dividend_fee in dividend_transaction.fees:
                    if dividend_fee.currency == currency:
                        total_fees_amount += dividend_fee.amount
            
            total_tax_amount = 0
            for dividend_transaction in dividend_transactions:
                for dividend_tax in dividend_transaction.taxes:
                    if dividend_tax.currency == currency:
                        total_tax_amount += dividend_tax.amount
                        
                        
            dividend_transaction = dividend_transactions[0]
            country_domiciled = dividend_transaction.asset.country_domiciled
            
            dividend = {
                "Ticker": dividend_transaction.asset.ticker,
                "Código Rendimento": "E11 - Dividendos ou lucros - sem retenção em Portugal",
                "País da fonte": f"{country_domiciled.code} - {country_domiciled.name_pt}",
                "Rendimento Bruto": total_net_amount,
                "Imposto Pago no Estrangeiro - No país da fonte": total_fees_amount + total_tax_amount
            }
            
            
            dividends.append(dividend)
        
        return dividends

File: src/formatters/pt/irs/dividends_formatter.ts
import { Currency } from "../../../models/currency.js";
import { Dividend, DividendForUser } from "../../../models/dividend.js";
import { Statement } from "../../../models/statement.js";
import { Transaction } from "../../../models/transaction.js";

type Dividends = {
  toUser: DividendForUser[];
  toIRS: Dividend[];
};

class PTDividendsFormatter {
  constructor() {}

  async format(
    statement: Statement,
    year?: number,
    currency: string = "EUR"
  ): Promise<Dividends> {
    // Obter transacoes referentes aos dividendos do ano a declarar
    const transactions = statement.getTransactions();
    const dividendTransactions = transactions.filter(
      (t) =>
        t.type === "Dividend" && (!year || t.date.startsWith(year.toString()))
    );
    const dividends = [];
    const dividendsByAsset: Record<string, Record<string, Transaction[]>> = {};

    // Agrupar dividendos por ativo e ano
    dividendTransactions.forEach((transaction) => {
      if (!dividendsByAsset[transaction.asset.isin]) {
        dividendsByAsset[transaction.asset.isin] = {};
      }
      if (
        !dividendsByAsset[transaction.asset.isin][
          transaction.date.substring(0, 4)
        ]
      ) {
        dividendsByAsset[transaction.asset.isin][
          transaction.date.substring(0, 4)
        ] = [];
      }
      dividendsByAsset[transaction.asset.isin][
        transaction.date.substring(0, 4)
      ].push(transaction);
    });
    const currencyConverter = new Currency();
    // console.log("dividendsByAsset", dividendsByAsset);
    let listDividendsByAsset = Object.values(dividendsByAsset);
    for (let listAssetDividends of listDividendsByAsset) {
      let listDividendsByAssetAndYear = Object.values(listAssetDividends);
      for (let listDividends of listDividendsByAssetAndYear) {
        let totalNetAmount = 0;
        let totalFeesAmount = 0;
        let totalTaxAmount = 0;
        let totalGrossAmount = 0;

        for (let transaction of listDividends) {
          const fees = transaction.fees;
          const taxes = transaction.taxes;

          for (let fee of fees) {
            if (fee.currency === currency) {
              totalFeesAmount += fee.amount;
            } else if (fee.exchangeRate) {
              totalFeesAmount += fee.amount * fee.exchangeRate;
            } else {
              totalFeesAmount += await currencyConverter.convert(
                fee.amount,
                fee.currency,
                currency,
                transaction.date
              );
            }
          }
          for (let tax of taxes) {
            if (tax.currency === currency) {
              totalTaxAmount += tax.amount;
            } else if (tax.exchangeRate) {
              totalTaxAmount += tax.amount * tax.exchangeRate;
            } else {
              totalTaxAmount += await currencyConverter.convert(
                tax.amount,
                tax.currency,
                currency,
                transaction.date
              );
            }
          }
          if (transaction.netAmountCurrency === currency) {
            totalNetAmount += transaction.netAmount;
          } else if (transaction.exchangeRate) {
            totalNetAmount += transaction.netAmount * transaction.exchangeRate;
          } else {
            totalNetAmount += await currencyConverter.convert(
              transaction.netAmount,
              transaction.netAmountCurrency,
              currency,
              transaction.date
            );
          }
        }
        // Calcular rendimento bruto
        // Devido às limitações em JavaScript de "float-point" errors, o valor do rendimento bruto deve ser arredondado para 2 casas decimais
        totalFeesAmount = parseFloat(totalFeesAmount.toFixed(2));
        totalTaxAmount = parseFloat(totalTaxAmount.toFixed(2));

        const totalExpenses = parseFloat(
          (totalFeesAmount + totalTaxAmount).toFixed(2)
        );
        totalGrossAmount = totalNetAmount + totalExpenses;
        totalGrossAmount = parseFloat(totalGrossAmount.toFixed(2));

        const dividendTransaction = listDividends[0];
        const countryDomiciled = dividendTransaction.asset.countryDomiciled;

        // if(dividendTransaction.asset.ticker == "MDT") {
        //   console.log("totalFeesAmount", totalFeesAmount);
        //   console.log("totaltaxamount", totalTaxAmount);
        // }

        let entry: DividendForUser = {
          Ticker: dividendTransaction.asset.ticker,
          "Ano rendimento": dividendTransaction.date.substring(0, 4),
          "Código Rendimento":
            "E11 - Dividendos ou lucros - sem retenção em Portugal",
          "País da fonte": `${countryDomiciled?.code} - ${countryDomiciled?.namePt}`,
          "Rendimento Bruto": totalGrossAmount,
          "Imposto Pago no Estrangeiro - No país da fonte": totalExpenses,
        };

        dividends.push(entry);
      }
    }

    // Como na AT a chave de cada linha é o país da fonte e o código de rendimento, temos de agrupar segundo isso
    // Se quiseres ignorar este facto para poderes mostrar os dividendos por ativo, comenta o seguinte bloco de código
    const dividendsForUser = dividends;
    let dividendsForIRS = dividends.reduce((acc: Dividend[], curr) => {
      let ref = acc.find(
        (dividend) =>
          dividend["Código Rendimento"] === curr["Código Rendimento"] &&
          dividend["País da fonte"] === curr["País da fonte"] &&
          dividend["Ano rendimento"] === curr["Ano rendimento"]
      );
      if (ref) {
        ref["Rendimento Bruto"] += curr["Rendimento Bruto"];
        ref["Imposto Pago no Estrangeiro - No país da fonte"] +=
          curr["Imposto Pago no Estrangeiro - No país da fonte"];
      } else {
        const { Ticker, ...rest } = curr;
        acc.push(rest);
      }
      return acc;
    }, []);

    return {
      toUser: dividendsForUser,
      toIRS: dividendsForIRS,
    };
  }
}

export { PTDividendsFormatter };
